File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys

input = sys.stdin.readline

N = int(input())

# 포도주를 선택하면 무조건 마셔야함
# 연속으로 놓여 있는 3잔을 모두 먹을 수 없음

wines = [0]
dp = [0] * 10001

for _ in range(N):
    wines.append(int(input()))

if N >= 1:
    dp[1] = wines[1]
if N >= 2:
    dp[2] = wines[1] + wines[2]

for i in range(3, N + 1):
    # 1. i번째 와인을 마시지 않는 경우 그냥 그전의 최고값을 쓰기
    # 2. dp[i-3] + wines[i-1] + wines[i] 즉 연속으로 두잔을 선택하는 경우의수
    # 3. dp[i-2] + wines[i] 한잔 을 뛰어넘어서 선택하는 경우의수
    dp[i] = max(dp[i - 1], dp[i - 3] + wines[i - 1] + wines[i], dp[i - 2] + wines[i])

print(max(dp))

File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] wines = new int[n + 1];
        for (int i = 1; i < n + 1; i++) {
            st = new StringTokenizer(br.readLine());
            wines[i] = Integer.parseInt(st.nextToken());
        }
        int[] dp = new int[n + 1];
        int answer = 0;
        if (n >= 1) {
            dp[1] = wines[1];
            answer = dp[1];
        }
        if (n >= 2) {
            dp[2] = wines[1] + wines[2];
            answer = dp[2];
        }

        for (int i = 3; i < n + 1; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + wines[i]);
            dp[i] = Math.max(dp[i], dp[i - 3] + wines[i - 1] + wines[i]);
            if (dp[i] > answer) {
                answer = dp[i];
            }
        }
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(String.valueOf(answer));
        bw.flush();
        bw.close();
    }
}